README.md
Ambari Cookbook
===============
![Travis CI status](https://travis-ci.org/jp/ambari.svg)

Install ambari-server and ambari-agent in order to provision a cluster with the Hortonworks Data Platform.  This cookbook also supports provisioning supported services through the Ambari API via Ambari Blueprints.  For more on blueprints, please refer to the Apache project's documentation [here](https://cwiki.apache.org/confluence/display/AMBARI/Blueprints).

Supports Ambari 1.7, 2.0,, 2.1 and 2.2.

Requirements
------------
Supports all OS distributions and versions supported by Ambari.

- RHEL, CentOS, and Oracle Linux 5, 6, and 7.
- Suse 11
- Ubuntu 12 and 14.
- Debian 7

Usage
-----
#### ambari::server

Include `ambari::server` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[ambari::server]"
  ]
}
```

#### ambari::agent

Include `ambari::agent` in your node's `run_list` (all the nodes of your cluster):

The Ambari Server will be automatically detected, but you can override this detection using the attribute `node['ambari']['server_fqdn']`.

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[ambari::agent]"
  ]
}
```

#### ambari::blueprints

Include `ambari::blueprints` in your node's `run_list`. Then configure the `blueprints` attributes listed below.

Main Attributes
----------

<table>
  <tr>
    <th>Key</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>node['ambari']['rhel_5_repo']</tt></td>
    <td>URL for the RHEL 5.x repository</td>
    <td><tt>http://public-repo-1.hortonworks.com/ambari/centos5/2.x/updates/2.0.0/ambari.repo</tt></td>
  </tr>
  <tr>
    <td><tt>node['ambari']['rhel_6_repo']</tt></td>
    <td>URL for the RHEL 6.x repository</td>
    <td><tt>http://public-repo-1.hortonworks.com/ambari/centos6/2.x/updates/2.0.0/ambari.repo</tt></td>
  </tr>
  <tr>
    <td><tt>node['ambari']['suse_11_repo']</tt></td>
    <td>URL for the Suse 11.x repository</td>
    <td><tt>http://public-repo-1.hortonworks.com/ambari/suse11/2.x/updates/2.0.0/ambari.repo</tt></td>
  </tr>
  <tr>
    <td><tt>node['ambari']['server_fdqn']</tt></td>
    <td>FQDN for the Ambari server</td>
    <td><tt>Found by searching a node using the recipe ambari::server in the same environment</tt></td>
  </tr>
</table>

Blueprints Attributes
----------

<table>
  <tr>
    <th>Key</th>
    <th>Description</th>
  </tr>
  <tr>
    <td><tt>node['ambari']['blueprints']['blueprint_name']</tt></td>
    <td>Name of the cluster's blueprint</td>
  </tr>
  <tr>
    <td><tt>node['ambari']['blueprints']['blueprint_json']</tt></td>
    <td>JSON description of the Blueprint. See a single-node example here : https://cwiki.apache.org/confluence/display/AMBARI/Blueprints#Blueprints-ExampleBlueprint</td>
  </tr>
  <tr>
    <td><tt>node['ambari']['blueprints']['cluster_name']</tt></td>
    <td>Cluster's name</td>
  </tr>
  <tr>
    <td><tt>node['ambari']['blueprints']['cluster_json']</tt></td>
    <td>JSON template for the Cluster. See a single-node example here: https://cwiki.apache.org/confluence/display/AMBARI/Blueprints#Blueprints-ExampleClusterCreationTemplate</td>
  </tr>
</table>


Contributing
------------

If you wish to contribute on this cookbook:

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Author: [Julien Pellet](https://twitter.com/julienpellet)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
ambari CHANGELOG
================

This file is used to list changes made in each version of the ambari cookbook.

0.2.4
-----
- [AltJ] - Support for Ambari 2.2
- [dsklopp] - Support for Ubuntu 14.04


recipes/agent.rb
#
# Cookbook Name:: ambari
# Recipe:: agent
#
# Copyright 2014, JULIEN PELLET
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'ambari::default'
include_recipe 'ambari::setup_package_manager'

package 'ambari-agent'

directory '/etc/ambari-agent/conf.chef' do
  owner 'root'
  group 'root'
  mode '0755'
  action :create
end

execute 'move existing ambari configuration' do
  command 'mv /etc/ambari-agent/conf /etc/ambari-agent/conf.dist'
  only_if do
    !::File.exist?('/etc/ambari-agent/conf.dist')
  end
end

execute 'alternatives configured confdir' do
  command 'update-alternatives --install /etc/ambari-agent/conf ambari-agent-conf /etc/ambari-agent/conf.chef 90'
  not_if 'update-alternatives --display ambari-agent-conf |grep "/etc/ambari-agent/conf.chef"'
end

# Get ambari-server fqdn
if Chef::Config[:solo]
  errmsg = 'This recipe uses search, Chef Solo does not support search.'
  Chef::Log.warn(errmsg)
elseif tagged?('ambari')
  ambari_server_fqdn = node['fqdn']
else
  ambari_server_fqdn = search('node', "tags:ambari AND chef_environment:#{node.chef_environment}").first['fqdn']
end

template '/etc/ambari-agent/conf/ambari-agent.ini' do
  source 'ambari-agent.ini.erb'
  mode 0755
  user 'root'
  group 'root'
  variables(ambari_server_fqdn: ambari_server_fqdn)
end

service 'ambari-agent' do
  supports :status => true, :restart => true, :reload => false
  action [:enable, :start]
end


recipes/server.rb
#
# Cookbook Name:: ambari
# Recipe:: server
#
# Copyright 2014, JULIEN PELLET
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'ambari::default'
include_recipe 'ambari::setup_package_manager'

package 'ambari-server'

# This could probably be a bit more clean instead of having a nested case.
# it works now, but will be ugly if we add support for all cases.
case node['ambari']['database']['type']
when 'embedded'
  db_opts = ''
when 'mysql'
  db_opts = "--database=#{node['ambari']['database']['type']} \
    --databaseport=#{node['ambari']['database']['port']} \
    --databasehost=#{node['ambari']['database']['host']} \
    --databasename=#{node['ambari']['database']['name']} \
    --databaseusername=#{node['ambari']['database']['username']} \
    --databasepassword=#{node['ambari']['database']['password']}"
when 'mssql', 'oracle', 'postgres', 'sqlanywhere'
  # needs db_opts cause the options differ from each db type.
  raise "#{node['ambari']['database']['type']} is not supported yet"
end

# this case sets the jdbc driver package name for each distro and DB type (unless database type is 'embedded').
case node['platform']
when 'ubuntu', 'debian'
  jdbcpkg = 'libmysql-java' if node['ambari']['database']['type'] == 'mysql'
when 'redhat', 'centos', 'amazon', 'scientific'
  # basically needs to know the "mysql java connector" package name for each distro, like the ubuntu/debian above.
  raise "#{node['ambari']['database']['type']} is not supported yet for RPM based distributions"
  jdbcpkg = 'XXXXXXXXXXXX' if node['ambari']['database']['type'] == 'mysql'
when 'suse'
  # basically needs to know the "mysql java connector" package name for each distro, like the ubuntu/debian above.
  raise "#{node['ambari']['database']['type']} is not supported yet for RPM based distributions"
  jdbcpkg = 'XXXXXXXXXXXX' if node['ambari']['database']['type'] == 'mysql'
end unless node['ambari']['database']['type'] == 'embedded'

# install jdbc driver.
if node['ambari']['jdbc']['url'] == ''
  package jdbcpkg
else
  remote_file node['ambari']['jdbc']['path'] do
    source node['ambari']['jdbc']['url']
    not_if { ::File.exist?(node['ambari']['jdbc']['path']) }
  end
end

execute 'setup ambari-server' do
  command "ambari-server setup #{db_opts} -s && touch /etc/ambari-server/.configured"
  creates '/etc/ambari-server/.configured'
end
  
if node['ambari']['database']['type'] == 'embedded'
  service 'postgresql' do
    supports :status => true, :restart => true, :reload => true
    action [:enable, :start]
  end
else
  service 'postgresql' do
    action [:disable, :stop]
  end
end

service 'ambari-server' do
  supports :status => true, :restart => true, :reload => false
  action [:start, :enable]
end

tag('ambari')


recipes/blueprints.rb
#
# Cookbook Name:: ambari
# Recipe:: blueprints
#
# Copyright 2014, JULIEN PELLET
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

package 'curl'

if Chef::Config[:solo]
  Chef::Log.warn('This recipe uses search. Chef Solo does not support search.')
else
  ambari_server_fqdn = node['ambari']['server_fqdn'] || search('node', 'run_list:recipe\[ambari\:\:server\] AND chef_environment:' + node.chef_environment).first['fqdn']
end
basic_auth_parameters = "--user #{node['ambari']['admin_user']}:#{node['ambari']['admin_password']}"

execute 'Init Blueprints' do
  command "curl #{basic_auth_parameters} -H 'X-Requested-By:ambari-cookbook' --data '#{node['ambari']['blueprints']['blueprint_json'].to_json}' #{ambari_server_fqdn}:8080/api/v1/blueprints/#{node['ambari']['blueprints']['blueprint_name']}"
end

execute 'Init Cluster' do
  command "curl #{basic_auth_parameters} -H 'X-Requested-By:ambari-cookbook' --data '#{node['ambari']['blueprints']['cluster_json'].to_json}' #{ambari_server_fqdn}:8080/api/v1/clusters/#{node['ambari']['blueprints']['cluster_name']}"
end


recipes/setup_package_manager.rb
#
# Cookbook Name:: ambari
# Recipe:: setup_package_manager
#
# Copyright 2014, JULIEN PELLET
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

package 'wget' # not sure if this is needed for remote_file, lets have it anyway.

case node['platform']
when 'redhat', 'centos', 'amazon', 'scientific'
  case node['platform_version'].to_i
  when 5
    yum_repo = node['ambari']['rhel_5_repo']
  when 6
    yum_repo = node['ambari']['rhel_6_repo']
  when 7
    yum_repo = node['ambari']['rhel_7_repo']
  end
  remote_file '/etc/yum.repos.d/ambari.repo' do
    source yum_repo
    not_if { ::File.exist?('/etc/yum.repos.d/ambari.repo') }
  end
when 'suse'
  remote_file '/etc/zypp/repos.d/ambari.repo' do
    source node['ambari']['suse_11_repo']
    not_if { ::File.exist?('/etc/zypp/repos.d/ambari.repo') }
  end
when 'ubuntu'
  include_recipe 'apt'

  case node['platform_version']
  when '12.04'
    apt_repository 'ambari' do
      uri node['ambari']['ubuntu_12_repo']
      distribution 'Ambari'
      components ['main']
      keyserver    'hkp://keyserver.ubuntu.com:80'
      key          'B9733A7A07513CAD'
    end
  when '14.04'
    apt_repository 'ambari' do
      uri node['ambari']['ubuntu_14_repo']
      distribution 'Ambari'
      components ['main']
      keyserver    'hkp://keyserver.ubuntu.com:80'
      key          'B9733A7A07513CAD'
    end
  end
end


attributes/default.rb
#
# Cookbook Name:: ambari
# Attributes:: default
#

# default['ambari']['server_fqdn'] = 'localhost'

# Set default version of Ambari.  Overwrite this to install a different version.
default['ambari']['version'] = '2.2'

case node['ambari']['version']
when '1.7'
  default['ambari']['rhel_5_repo'] = 'http://public-repo-1.hortonworks.com/ambari/centos5/1.x/updates/1.7.0/ambari.repo'
  default['ambari']['rhel_6_repo'] = 'http://public-repo-1.hortonworks.com/ambari/centos6/1.x/updates/1.7.0/ambari.repo'
  default['ambari']['suse_11_repo'] = 'http://public-repo-1.hortonworks.com/ambari/suse11/1.x/updates/1.7.0/ambari.repo'
  default['ambari']['ubuntu_12_repo'] = 'http://public-repo-1.hortonworks.com/ambari/ubuntu12/1.x/updates/1.7.0'
when '2.0', '2'
  default['ambari']['rhel_5_repo'] = 'http://public-repo-1.hortonworks.com/ambari/centos5/2.x/updates/2.0.0/ambari.repo'
  default['ambari']['rhel_6_repo'] = 'http://public-repo-1.hortonworks.com/ambari/centos6/2.x/updates/2.0.0/ambari.repo'
  default['ambari']['suse_11_repo'] = 'http://public-repo-1.hortonworks.com/ambari/suse11/2.x/updates/2.0.0/ambari.repo'
  default['ambari']['ubuntu_12_repo'] = 'http://public-repo-1.hortonworks.com/ambari/ubuntu12/2.x/updates/2.0.0'
when '2.0.1'
  default['ambari']['rhel_5_repo'] = 'http://public-repo-1.hortonworks.com/ambari/centos5/2.x/updates/2.0.1/ambari.repo'
  default['ambari']['rhel_6_repo'] = 'http://public-repo-1.hortonworks.com/ambari/centos6/2.x/updates/2.0.1/ambari.repo'
  default['ambari']['suse_11_repo'] = 'http://public-repo-1.hortonworks.com/ambari/suse11/2.x/updates/2.0.1/ambari.repo'
  default['ambari']['ubuntu_12_repo'] = 'http://public-repo-1.hortonworks.com/ambari/ubuntu12/2.x/updates/2.0.1'
when '2.1'
  # RHEL 5 no longer supported in Ambari 2.1
  default['ambari']['rhel_6_repo'] = 'http://public-repo-1.hortonworks.com/ambari/centos6/2.x/updates/2.1.2/ambari.repo'
  default['ambari']['rhel_7_repo'] = 'http://public-repo-1.hortonworks.com/ambari/centos7/2.x/updates/2.1.2/ambari.repo'
  default['ambari']['suse_11_repo'] = 'http://public-repo-1.hortonworks.com/ambari/suse11/2.x/updates/2.1.2/ambari.repo'
  default['ambari']['ubuntu_12_repo'] = 'http://public-repo-1.hortonworks.com/ambari/ubuntu12/2.x/updates/2.1.2.1'
  default['ambari']['ubuntu_14_repo'] = 'http://public-repo-1.hortonworks.com/ambari/ubuntu14/2.x/updates/2.1.2.1'
  default['ambari']['debian_7_repo'] = 'http://public-repo-1.hortonworks.com/ambari/debian7/2.x/updates/2.1.2'
when '2.2'
  default['ambari']['rhel_6_repo'] = 'http://public-repo-1.hortonworks.com/ambari/centos6/2.x/updates/2.2.0.0/ambari.repo'
  default['ambari']['rhel_7_repo'] = 'http://public-repo-1.hortonworks.com/ambari/centos7/2.x/updates/2.2.0.0/ambari.repo'
  default['ambari']['suse_11_repo'] = 'http://public-repo-1.hortonworks.com/ambari/suse11/2.x/updates/2.2.0.0/ambari.repo'
  default['ambari']['ubuntu_12_repo'] = 'http://public-repo-1.hortonworks.com/ambari/ubuntu12/2.x/updates/2.2.0.0'
  default['ambari']['ubuntu_14_repo'] = 'http://public-repo-1.hortonworks.com/ambari/ubuntu14/2.x/updates/2.2.0.0'
  default['ambari']['debian_7_repo'] = 'http://public-repo-1.hortonworks.com/ambari/debian7/2.x/updates/2.2.0.0'
end

default['ambari']['admin_user'] = 'admin'
default['ambari']['admin_password'] = 'admin'

# DB attributes
default['ambari']['database']['type'] = 'embedded' # embedded|mysql (can support following in the future oracle|mssql|postgres|sqlanywhere)
# These attributes below are omittied if database type is 'embedded'.
default['ambari']['database']['host'] = 'localhost'
default['ambari']['database']['port'] = '3306'
default['ambari']['database']['name'] = 'ambari'
default['ambari']['database']['username'] = 'ambari'
default['ambari']['database']['password'] = 'bigdata'

# JDBC attributes, will only be used if url is specified and db type isn't 'embedded'.
default['ambari']['jdbc']['url'] = '' # leave blank if no remote file
default['ambari']['jdbc']['path'] = '/usr/share/java/mysql-connector-java.jar'
